File: src/main/java/org/kite9/diagram/visualization/planarization/rhd/layout/ExitMatrixEvaluator.java
package org.kite9.diagram.visualization.planarization.rhd.layout;

import org.kite9.diagram.common.elements.RoutingInfo;
import org.kite9.diagram.common.objects.Bounds;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase.Group;
import org.kite9.diagram.visualization.planarization.rhd.layout.ExitMatrix.RelativeSide;
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkDetail;
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkProcessor;
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D;
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D.DPos;

/**
 * Tries to work out the likelihood of an overlap.
 * 
 * @author robmoffat
 * 
 */
public class ExitMatrixEvaluator {

	private static final float OCCLUDED_COST = .5f;
	private static final float STRAIGHT_VS_CORNER_COST = .5f;

	/**
	 * A heuristic calculation for the likely number of crossings due to this arragement
	 */
	public static double countOverlaps(ExitMatrix aMatrix, ExitMatrix bMatrix, Layout aDirection, RoutableHandler2D rh) {
		float sideA = sumOneSide(aMatrix, bMatrix, aDirection, -1);
		float sideB = sumOneSide(aMatrix, bMatrix, aDirection, +1);
		float occlusionA = addOneOccludedSide(aMatrix, bMatrix, aDirection, rh);
		float occlusionB = addOneOccludedSide(bMatrix, aMatrix, Layout.reverse(aDirection), rh);

		return sideA + sideB + occlusionA + occlusionB;
	}
	
	/**
	 * Works out the extra distance of external links due to this arrangement
	 */
	public static double calculateExtraExternalLinkDistance(ExitMatrix aMatrix, ExitMatrix bMatrix, Layout aDirection, RoutableHandler2D rh) {
		return calcEEDistanceOneSide(aMatrix, bMatrix.getSizeInDirectionOfLayout(aDirection), aDirection) + 
			calcEEDistanceOneSide(bMatrix, aMatrix.getSizeInDirectionOfLayout(aDirection), Layout.reverse(aDirection)); 
	}
	
	
	

	private static double calcEEDistanceOneSide(ExitMatrix aMatrix, Bounds distance, Layout d) {
		float count = aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, -1)+
			aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, 0)+
			aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, 1);
		
		double incrementalDistance = distance.getDistanceMax() - distance.getDistanceMin();
		
		return count * incrementalDistance;
	}

	/**
	 * Works out the cost score for the distance travelled by the internal links
	 */
	public static double calculateInternalDistance(LinkDetail internal, final RoutableHandler2D rh) {
		if (internal == null)
			return 0;
		
		final double out[] = { 0 };
		internal.processLowestLevel(new LinkProcessor() {
			
			@Override
			public void process(Group from, Group to, LinkDetail ld) {
				RoutingInfo aRI = from.getAxis().getPosition(rh, true);
				RoutingInfo bRI = to.getAxis().getPosition(rh, true);
				double cost = rh.cost(aRI, bRI) * ld.getNumberOfLinks();
				out[0] += cost;
//				log.send("Evaluating: "+cost+
//						"\n\tfrom "+((LeafGroup)from).getContained()+" at "+aRI+
//						"\n\tto "+((LeafGroup)to).getContained()+" at "+bRI);
			}
		});
		
		return out[0];
	}

	/**
	 * Figures out a cost associated with one side of the matrix based on the
	 * likelihood of leaving edges crossing one another.
	 */
	private static float sumOneSide(ExitMatrix aMatrix, ExitMatrix bMatrix, Layout ad, int rank) {
		float bFacing = bMatrix.getLinkCount(ad, RelativeSide.FACING, rank);
		float aMid = aMatrix.getLinkCount(ad, RelativeSide.MIDDLE, rank);
		float bMid = bMatrix.getLinkCount(ad, RelativeSide.MIDDLE, rank);
		float aOpp = aMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, rank);

		float cornerCross = bFacing * aOpp;
		float midCrossB = (aOpp * bMid) * STRAIGHT_VS_CORNER_COST;
		float midCrossA = (bFacing * aMid) * STRAIGHT_VS_CORNER_COST;

		return cornerCross + midCrossB + midCrossA;
	}

	/**
	 * Work out how many edges you will have to cross to get from the occluded square to where you want to go.
	 * 
	 * When a matrix with a small span is occluded by one with a large span, we need to 
	 * add crossings in order to "break in" to the space of the larger one.
	 */
	private static float addOneOccludedSide(ExitMatrix aMatrix, final ExitMatrix bMatrix, Layout ad,
			RoutableHandler2D rh) {
		float aOccluded = aMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, 0); 
		
		if (aOccluded == 0) {
			return 0;
		}

		
		float facingOccluding = bMatrix.getLinkCount(ad, RelativeSide.FACING, 0) / 2;
		float lowerOccluding = getLinkCountOnSide(bMatrix, ad, -1) + facingOccluding;
		float higherOccluding = getLinkCountOnSide(bMatrix, ad, 1)+ facingOccluding;
		
		lowerOccluding *= aOccluded;
		higherOccluding *= aOccluded;
		
		// large span check
		Bounds aSpan = aMatrix.getSpanInDirectionOfLayout(Layout.reverse(ad));
		Bounds bSpan = bMatrix.getSpanInDirectionOfLayout(Layout.reverse(ad));
		
		if (rh.compareBounds(aSpan, bSpan) == DPos.OVERLAP) {
			float bOccluding = bMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, 0);
			
			// works out the occluded span
			double occludedMin = Math.max(aSpan.getDistanceMin(), bSpan.getDistanceMin());
			double occludedMax = Math.min(aSpan.getDistanceMax(), bSpan.getDistanceMax());
			double topOccludedFrac = ( occludedMax - occludedMin ) / ((aSpan.getDistanceMax() - aSpan.getDistanceMin()));
			
			// work out how much of b you would have to cross to get back to where you started from
			double lowerTravel = occludedMax - bSpan.getDistanceMin();
			double higherTravel = bSpan.getDistanceMax() - occludedMin;
				
			// work out travel as a fraction of b
			double lowerFrac = lowerTravel / (bSpan.getDistanceMax() - bSpan.getDistanceMin());
			double higherFrac = higherTravel / (bSpan.getDistanceMax() - bSpan.getDistanceMin());
			
			// work out how many edge crossings that's likely to be
			double lowerAmt = topOccludedFrac * lowerFrac * bOccluding * aOccluded;
			double higherAmt = topOccludedFrac * higherFrac * bOccluding* aOccluded;
			lowerOccluding += lowerAmt;
			higherOccluding += higherAmt;
		}
		
		return Math.min(lowerOccluding, higherOccluding) * OCCLUDED_COST;
	}

	private static float getLinkCountOnSide(final ExitMatrix mat, Layout ad, int side) {
		return mat.getLinkCount(ad, RelativeSide.FACING, side) + mat.getLinkCount(ad, RelativeSide.MIDDLE, side)
				+ mat.getLinkCount(ad, RelativeSide.OPPOSITE, side);
	}

}

File: kite9-visualization/src/commonMain/kotlin/org/kite9/diagram/visualization/planarization/rhd/layout/ExitMatrixEvaluator.kt
package org.kite9.diagram.visualization.planarization.rhd.layout

import kotlin.math.max
import kotlin.math.min
import org.kite9.diagram.common.objects.Bounds
import org.kite9.diagram.model.position.Layout
import org.kite9.diagram.model.position.Layout.Companion.reverse
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.group.Group
import org.kite9.diagram.visualization.planarization.rhd.layout.ExitMatrix.RelativeSide
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkDetail
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkProcessor
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D.DPos

/**
 * Tries to work out the likelihood of an overlap.
 *
 * @author robmoffat
 */
class ExitMatrixEvaluator {

    /** A heuristic calculation for the likely number of crossings due to this arragement */
    fun countOverlaps(
            aMatrix: ExitMatrix,
            bMatrix: ExitMatrix,
            aDirection: Layout,
            rh: RoutableHandler2D
    ): Double {
        val sideA = sumOneSide(aMatrix, bMatrix, aDirection, -1)
        val sideB = sumOneSide(aMatrix, bMatrix, aDirection, +1)
        val occlusionA = addOneOccludedSide(aMatrix, bMatrix, aDirection, rh)
        val occlusionB = addOneOccludedSide(bMatrix, aMatrix, reverse(aDirection), rh)
        return (sideA + sideB + occlusionA + occlusionB).toDouble()
    }

    /** Works out the extra distance of external links due to this arrangement */
    fun calculateExtraExternalLinkDistance(
            aMatrix: ExitMatrix,
            bMatrix: ExitMatrix,
            aDirection: Layout?,
            rh: RoutableHandler2D?
    ): Double {
        return calcEEDistanceOneSide(
                aMatrix,
                bMatrix.getSizeInDirectionOfLayout(aDirection),
                aDirection
        ) +
                calcEEDistanceOneSide(
                        bMatrix,
                        aMatrix.getSizeInDirectionOfLayout(aDirection),
                        reverse(aDirection)
                )
    }

    private fun calcEEDistanceOneSide(aMatrix: ExitMatrix, distance: Bounds, d: Layout?): Double {
        val count =
                aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, -1) +
                        aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, 0) +
                        aMatrix.getLinkCount(d, RelativeSide.OPPOSITE, 1)
        val incrementalDistance = distance.distanceMax - distance.distanceMin
        return count * incrementalDistance
    }

    /** Works out the cost score for the distance travelled by the internal links */
    fun calculateInternalDistance(internal: LinkDetail?, rh: RoutableHandler2D): Double {
        if (internal == null) return 0.0
        val out = doubleArrayOf(0.0)
        internal.processLowestLevel(
                object : LinkProcessor {
                    override fun process(
                            originatingGroup: Group,
                            destinationGroup: Group,
                            ld: LinkDetail
                    ) {
                        val aRI = originatingGroup.axis.getPosition(rh, true)
                        val bRI = destinationGroup.axis.getPosition(rh, true)
                        val cost = rh.cost(aRI, bRI) * ld!!.numberOfLinks
                        out[0] += cost
                        //				log.send("Evaluating: "+cost+
                        //						"\n\tfrom "+((LeafGroup)originatingGroup).getContained()+" at "+aRI+
                        //						"\n\tto "+((LeafGroup)destinationGroup).getContained()+" at "+bRI);
                    }
                }
        )
        return out[0]
    }

    /**
     * Figures out a cost associated with one side of the matrix based on the likelihood of leaving
     * edges crossing one another.
     */
    private fun sumOneSide(aMatrix: ExitMatrix, bMatrix: ExitMatrix, ad: Layout, rank: Int): Float {
        val bFacing = bMatrix.getLinkCount(ad, RelativeSide.FACING, rank)
        val aMid = aMatrix.getLinkCount(ad, RelativeSide.MIDDLE, rank)
        val bMid = bMatrix.getLinkCount(ad, RelativeSide.MIDDLE, rank)
        val aOpp = aMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, rank)
        val cornerCross = bFacing * aOpp
        val midCrossB = aOpp * bMid * STRAIGHT_VS_CORNER_COST
        val midCrossA = bFacing * aMid * STRAIGHT_VS_CORNER_COST
        return cornerCross + midCrossB + midCrossA
    }

    /**
     * Work out how many edges you will have to cross to get from the occluded square to where you
     * want to go.
     *
     * When a matrix with a small span is occluded by one with a large span, we need to add
     * crossings in order to "break in" to the space of the larger one.
     */
    private fun addOneOccludedSide(
            aMatrix: ExitMatrix,
            bMatrix: ExitMatrix,
            ad: Layout?,
            rh: RoutableHandler2D
    ): Float {
        val aOccluded = aMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, 0)
        if (aOccluded == 0f) {
            return 0F
        }
        val facingOccluding = bMatrix.getLinkCount(ad, RelativeSide.FACING, 0) / 2
        var lowerOccluding = getLinkCountOnSide(bMatrix, ad, -1) + facingOccluding
        var higherOccluding = getLinkCountOnSide(bMatrix, ad, 1) + facingOccluding
        lowerOccluding *= aOccluded
        higherOccluding *= aOccluded

        // large span check
        val aSpan = aMatrix.getSpanInDirectionOfLayout(reverse(ad)!!)
        val bSpan = bMatrix.getSpanInDirectionOfLayout(reverse(ad)!!)
        if (rh.compareBounds(aSpan!!, bSpan!!) === DPos.OVERLAP) {
            val bOccluding = bMatrix.getLinkCount(ad, RelativeSide.OPPOSITE, 0)

            // works out the occluded span
            val occludedMin = max(aSpan!!.distanceMin, bSpan!!.distanceMin)
            val occludedMax = min(aSpan.distanceMax, bSpan.distanceMax)
            val topOccludedFrac =
                    (occludedMax - occludedMin) / (aSpan.distanceMax - aSpan.distanceMin)

            // work out how much of b you would have to cross to get back to where you started from
            val lowerTravel = occludedMax - bSpan.distanceMin
            val higherTravel = bSpan.distanceMax - occludedMin

            // work out travel as a fraction of b
            val lowerFrac = lowerTravel / (bSpan.distanceMax - bSpan.distanceMin)
            val higherFrac = higherTravel / (bSpan.distanceMax - bSpan.distanceMin)

            // work out how many edge crossings that's likely to be
            val lowerAmt = topOccludedFrac * lowerFrac * bOccluding * aOccluded
            val higherAmt = topOccludedFrac * higherFrac * bOccluding * aOccluded
            lowerOccluding += lowerAmt.toFloat()
            higherOccluding += higherAmt.toFloat()
        }
        return min(lowerOccluding, higherOccluding) * OCCLUDED_COST
    }

    private fun getLinkCountOnSide(mat: ExitMatrix, ad: Layout?, side: Int): Float {
        return (mat.getLinkCount(ad, RelativeSide.FACING, side) +
                mat.getLinkCount(ad, RelativeSide.MIDDLE, side) +
                mat.getLinkCount(ad, RelativeSide.OPPOSITE, side))
    }

    companion object {

        private const val OCCLUDED_COST = .5f
        private const val STRAIGHT_VS_CORNER_COST = .5f
    }
}
